baseplate/lib/message_queue.py

```diff
@@ -121,6 +159,176 @@ def close(self) -> None:
+        "message_queue_sidecar_latency",
+        "latency of message requests to the publishing sidecar.",
+        prom_labels,
+        buckets=default_latency_buckets,
+    )
+
+    def __init__(
+        self,
+        name: str,
+        host: str = DEFAULT_QUEUE_HOST,
+        port: int = DEFAULT_QUEUE_PORT,
+        pool_size: int = 10,
+        pool_timeout: int = 1,
+        pool_conn_max_age: int = 120,
+    ):
+        self.name = name
+        self.host = host
+        self.port = port
+        self.pool = self.create_connection_pool(pool_size, pool_timeout, pool_conn_max_age)
+
+    def create_connection_pool(
+        self, pool_size: int, pool_timeout: int, pool_conn_max_age: int
+    ) -> ThriftConnectionPool:
+        endpoint = config.Endpoint(f"{self.host}:{self.port}")
+        pool = ThriftConnectionPool(
+            endpoint, size=pool_size, timeout=pool_timeout, max_age=pool_conn_max_age
+        )
+        return pool
+
+    def _update_counters(self, outcome: str) -> None:
+        # This request is no longer queued
+        self.remote_queue_put_length.labels(
+            queue_name=self.name,
+        ).dec()
+        # Increment success/failure counters
+        if outcome == "success":
+            metric = self.remote_queue_put_success
+        else:
+            metric = self.remote_queue_put_fail
+        metric.labels(
+            queue_name=self.name,
+        ).inc()
+
+    def _put_success_callback(self, _: Any) -> None:
+        self._update_counters("success")
+
+    def _put_fail_callback(self, greenlet: Any) -> None:
+        self._update_counters("fail")
+        try:
+            greenlet.get()
+        except Exception as e:
+            logging.info("Remote queue `put` failed, exception found: %s", e)
+
+    def get(self, _: Optional[float] = None) -> bytes:
+        raise NotImplementedError  # This queue type is write-only
+
+    def _try_to_put(self, message: bytes, timeout: Optional[float], start_time: float) -> bool:
+        # get a connection from the pool
+        with self.pool.connection() as protocol:
+            client = RemoteMessageQueueService.Client(protocol)
+            try:
+                client.put(message, timeout)
+                # record latency
+                self.remote_queue_put_latency.labels(
+                    queue_name=self.name,
+                ).observe(time.perf_counter() - start_time)
+                return True  # Success
+            # If the server responded with a timeout, raise our own timeout to be consistent with the posix queue
+            except ThriftTimedOutError:
+                raise TimedOutError
+
+    def put(self, message: bytes, timeout: Optional[float] = None) -> Any:
+        # increment in-flight counter
+        self.remote_queue_put_length.labels(
+            queue_name=self.name,
+        ).inc()
+        start_time = time.perf_counter()
+
+        greenlet = gevent.spawn(self._try_to_put, message, timeout, start_time)
```

Comment:
small thing if gevent.spawn raises or fails for some reason, we won't decrement the counter. I think the easiest thing to do is to guard `gevent.spawn` around a `try: except` and decrement the counter if it fails. There's probably a better way to deal with that but I think that's fine for now